File: app/src/main/java/com/example/ruslanyussupov/popularmovies/detail/DetailViewModelFactory.java
package com.example.ruslanyussupov.popularmovies.detail;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.example.ruslanyussupov.popularmovies.data.model.Movie;

public class DetailViewModelFactory implements ViewModelProvider.Factory {

    private final Movie movie;

    public DetailViewModelFactory(Movie movie) {
        this.movie = movie;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(DetailViewModel.class)) {
            return (T) new DetailViewModel(movie);
        }
        throw new IllegalArgumentException("modelClass should be instance of DetailViewModel.class");
    }
}

File: app/src/main/java/com/example/ruslanyussupov/popularmovies/detail/DetailViewModelFactory.kt
package com.example.ruslanyussupov.popularmovies.detail

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

import com.example.ruslanyussupov.popularmovies.data.model.Movie

@Suppress("UNCHECKED_CAST")
class DetailViewModelFactory(private val movie: Movie) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(DetailViewModel::class.java)) {
            return DetailViewModel(movie) as T
        }
        throw IllegalArgumentException("modelClass should be instance of DetailViewModel.class")
    }
}
